File: app/src/main/java/com/dumuzeyn/mp3player/TabTransitionCoordinator.java
package com.dumuzeyn.mp3player;

import android.view.View;
import android.widget.ScrollView;

/** Commits tab navigation without keeping transition mechanics in the Activity. */
final class TabTransitionCoordinator {
    interface ContentBinding {
        void bind(ScrollView scrollView, android.widget.LinearLayout content);
    }

    interface Navigation {
        int selectedTab();

        void setSelectedTab(int value);

        String searchQuery();

        void setSearchQuery(String value);

        void setPreferredDirection(int value);

        void setTransitionRunning(boolean value);
    }

    private final Navigation navigation;
    private final MainScreenView screen;
    private final MainRenderer renderer;
    private final BackNavigationController backNavigation;
    private final TabsController tabs;
    private final SongsRenderer songs;
    private final ContentBinding contentBinding;

    TabTransitionCoordinator(Navigation navigation, MainScreenView screen,
            MainRenderer renderer, BackNavigationController backNavigation,
            TabsController tabs, SongsRenderer songs, ContentBinding contentBinding) {
        this.navigation = navigation;
        this.screen = screen;
        this.renderer = renderer;
        this.backNavigation = backNavigation;
        this.tabs = tabs;
        this.songs = songs;
        this.contentBinding = contentBinding;
    }

    void complete(TabTransitionRequest request) {
        recordHistoryIfNeeded(request);
        applyNavigation(request);
        renderer.render();
        tabs.finishTransition(request.targetTab);
    }

    void completeWithPreview(TabTransitionRequest request) {
        recordHistoryIfNeeded(request);
        renderer.captureCurrentScrollPosition();
        ScrollView previous = screen.contentScroll();
        if (previous != null && previous.getParent() == screen.contentHost()) {
            screen.contentHost().removeView(previous);
        }
        request.scrollView.setTranslationX(0.0f);
        request.scrollView.setAlpha(1.0f);
        request.scrollView.setVisibility(View.VISIBLE);
        request.scrollView.setOnScrollChangeListener(
                (view, scrollX, scrollY, oldScrollX, oldScrollY) ->
                        songs.loadMoreIfNearBottom());
        screen.replaceContent(request.scrollView, request.content);
        contentBinding.bind(request.scrollView, request.content);
        applyNavigation(request);
        renderer.adoptPreview(
                request.targetTab, request.searchQuery, request.previewState);
        tabs.refreshTabs();
        tabs.finishTransition(request.targetTab);
    }

    private void recordHistoryIfNeeded(TabTransitionRequest request) {
        if (request.recordHistory) {
            backNavigation.recordTabState(
                    navigation.selectedTab(), navigation.searchQuery());
        }
    }

    private void applyNavigation(TabTransitionRequest request) {
        navigation.setPreferredDirection(request.direction);
        navigation.setSelectedTab(request.targetTab);
        navigation.setSearchQuery(request.searchQuery);
        navigation.setTransitionRunning(false);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/TabTransitionCoordinator.kt
package com.dumuzeyn.mp3player

import android.view.View
import android.widget.LinearLayout
import android.widget.ScrollView

/** Commits tab navigation without keeping transition mechanics in the Activity. */
internal class TabTransitionCoordinator(
    private val navigation: Navigation,
    private val screen: MainScreenView,
    private val renderer: MainRenderer,
    private val backNavigation: BackNavigationController,
    private val tabs: TabsController,
    private val songs: SongsRenderer,
    private val contentBinding: ContentBinding,
) {
    fun interface ContentBinding {
        fun bind(scrollView: ScrollView, content: LinearLayout)
    }

    interface Navigation {
        fun selectedTab(): Int
        fun setSelectedTab(value: Int)
        fun searchQuery(): String
        fun setSearchQuery(value: String?)
        fun setPreferredDirection(value: Int)
        fun setTransitionRunning(value: Boolean)
    }

    fun complete(request: TabTransitionRequest) {
        recordHistoryIfNeeded(request)
        applyNavigation(request)
        renderer.render()
        tabs.finishTransition(request.targetTab)
    }

    fun completeWithPreview(request: TabTransitionRequest) {
        val previewScroll = requireNotNull(request.scrollView)
        val previewContent = requireNotNull(request.content)
        val previewState = requireNotNull(request.previewState)
        recordHistoryIfNeeded(request)
        renderer.captureCurrentScrollPosition()
        val previous = screen.contentScroll()
        val contentHost = screen.contentHost()
        if (previous != null && previous.parent === contentHost) {
            contentHost.removeView(previous)
        }
        previewScroll.translationX = 0f
        previewScroll.alpha = 1f
        previewScroll.visibility = View.VISIBLE
        previewScroll.setOnScrollChangeListener { _, _, _, _, _ -> songs.loadMoreIfNearBottom() }
        screen.replaceContent(previewScroll, previewContent)
        contentBinding.bind(previewScroll, previewContent)
        applyNavigation(request)
        renderer.adoptPreview(request.targetTab, request.searchQuery, previewState)
        tabs.refreshTabs()
        tabs.finishTransition(request.targetTab)
    }

    private fun recordHistoryIfNeeded(request: TabTransitionRequest) {
        if (request.recordHistory) {
            backNavigation.recordTabState(navigation.selectedTab(), navigation.searchQuery())
        }
    }

    private fun applyNavigation(request: TabTransitionRequest) {
        navigation.setPreferredDirection(request.direction)
        navigation.setSelectedTab(request.targetTab)
        navigation.setSearchQuery(request.searchQuery)
        navigation.setTransitionRunning(false)
    }
}
